worlds/kh1/Rules.py

```diff
@@ -0,0 +1,1962 @@
+    return (puppies_available >= puppies_required) or state.has("All Puppies", player)
+
+def has_torn_pages(state: CollectionState, player: int, pages_required: int) -> bool:
+    return state.count_from_list_unique(TORN_PAGES, player) >= pages_required
+
+def has_all_arts(state: CollectionState, player: int) -> bool:
+    return state.has_all({"Fire Arts", "Blizzard Arts", "Thunder Arts", "Cure Arts", "Gravity Arts", "Stop Arts", "Aero Arts"}, player)
+
+def has_all_summons(state: CollectionState, player: int) -> bool:
+    return state.has_all({"Simba", "Bambi", "Genie", "Dumbo", "Mushu", "Tinker Bell"}, player)
+
+def has_all_magic_lvx(state: CollectionState, player: int, level) -> bool:
+    return state.has_all_counts({
+        "Progressive Fire": level,
+        "Progressive Blizzard": level,
+        "Progressive Thunder": level,
+        "Progressive Cure": level,
+        "Progressive Gravity": level,
+        "Progressive Aero": level,
+        "Progressive Stop": level}, player)
+
+def has_offensive_magic(state: CollectionState, player: int) -> bool:
+    return state.has_any({"Progressive Fire", "Progressive Blizzard", "Progressive Thunder", "Progressive Gravity", "Progressive Stop"}, player)
+
+def has_reports(state: CollectionState, player: int, eotw_required_reports: int) -> bool:
+    return state.has_group_unique("Reports", player, eotw_required_reports)
+
+def has_final_rest_door(state: CollectionState, player: int, final_rest_door_requirement: str, final_rest_door_required_reports: int, keyblades_unlock_chests: bool):
+    if final_rest_door_requirement == "reports":
+        return state.has_group_unique("Reports", player, final_rest_door_required_reports)
+    if final_rest_door_requirement == "puppies":
+        return has_puppies(state, player, 99)
+    if final_rest_door_requirement == "postcards":
+        return state.has("Postcard", player, 10)
+    if final_rest_door_requirement == "superbosses":
+        return (
+                state.has_all({"Olympus Coliseum", "Neverland", "Agrabah", "Hollow Bastion", "Green Trinity", "Phil Cup", "Pegasus Cup", "Hercules Cup", "Entry Pass"}, player)
+                and has_emblems(state, player, keyblades_unlock_chests)
+                and has_all_magic_lvx(state, player, 2)
+                and has_defensive_tools(state, player)
+                and has_x_worlds(state, player, 7, keyblades_unlock_chests)
+            )
+
+def has_defensive_tools(state: CollectionState, player: int) -> bool:
+    return (
+            (
+                state.has("Progressive Cure", player, 2)
+                and state.has("Leaf Bracer", player)
+                and state.has("Dodge Roll", player)
+            )
+            and (
+                state.has("Second Chance", player)
+                or state.has("MP Rage", player)
+                or state.has("Progressive Aero", player, 2)
+            )
+        )
```

Comment:
This could be

```
return (
    state.has_all_counts({"Progressive Cure": 2, "Leaf Bracer": 1, "Dodge Roll": 1})
    and state.has_any_count({"Second Chance": 1, "MP Rage": 1, "Progressive Aero": 2})
)
```

*low priority / up to preference*